### PR TITLE
Improve fetch external docs workflow

### PR DESCRIPTION
**Pull request by jsetton**
* Improve the reusable fetch external docs workflow
* Add support for specifying the source repository ref (branch or tag)
* Add `openhab-android`, `openhab-google-assistant` and `openhab-mycroft` workflows
* Update `openhab-alexa` and `openhabian` workflows

Related to https://github.com/openhab/openhab-alexa/pull/457

### CODE REVIEW

**Comment by Confectrician** [comment]
Thanks Jeremy.

We could merge this one, but the workflow is not finished yet on the docs side anyway.

I have checked the PRs it produces and we will have to add some additional effort/logic to a repo structure like alexa has with a USAGE.md file.

We would have to rename the file to "readme.md" within the action.
But i don't want to touch this actoin a dozen times, each time a new repo gets added, so iw ill do some research about the different structures after christmas holidays for this.

Therefore i will leave this one open too for that time.

**Comment by jsetton** [comment]
> We would have to rename the file to "readme.md" within the action.

I confused as this is already part of the reusable fetch docs workflow.

https://github.com/openhab/openhab-docs/blob/61563aaface729ce1132f21a88661792282260af/.github/workflows/fetch_external_docs_reusable.yml#L69-L71

If your point is not wanting that specific action to be part of the workflow, I don't mind having the file rename in the source repository as readme.md as a requirement.

**Comment by Confectrician** [comment]
> I am confused as this is already part of the reusable fetch docs workflow.

That script part was a try that didn't work and i did not yet look into it. 🙂 

Generally I am fine with renaming the files on this side.
`USAGE.md` is common over several repositories.
(And i think USAGE is also a common naming in the github docs/help, so it is good to act close to the common github way too.)

**Comment by jsetton** [comment]
So I now understand what you mean. The working directory is missing from the rename file paths.

Anyway, I took a closer look at the current [pom.xml](https://github.com/openhab/openhab-docs/blob/final/pom.xml) file and noticed how different some of the external documentation integrations are (e.g. openhabian, openhab-android and openhab-alexa). It might actually be easier to have separate workflows with their own custom jobs. I think trying to build a reusable workflow template that works with all these integrations will require to make changes to some of them.

Here is how I would update the fetch `openhab-alexa` doc workflow:

```yml
name: Fetch OpenHAB Skill For Amazon Alexa Docs

on:
  # Repository dispatch event, to be triggerd by an openhab-alexa release
  # or manually from the openhab-alexa repository
  repository_dispatch:
    types: [update-openhab-alexa-docs-event]
  workflow_dispatch:

jobs:
  fetchDocs:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout repository
        uses: actions/checkout@v2.4.0
        with:
          ref: final-stable

      - name: Checkout external repository
        uses: actions/checkout@v2.4.0
        with:
          repository: openhab/openhab-alexa
          path: openhab-alexa
          ref: ${{ github.event.client_payload.ref }}

      - name: Update doc content
        run: |
          find _addons_ios/alexa-skill/ -type f -delete
          cp -v openhab-alexa/docs/images/*.png _addons_ios/alexa-skill/images
          cp -v openhab-alexa/docs/USAGE.md _addons_ios/alexa-skill/readme.md
          rm -rf openhab-alexa

      - name: Create pull request
        uses: peter-evans/create-pull-request@v3
        with:
          commit-message: Automated openhab-alexa Docs fetch
          committer: openHAB Bot <info@openhab.org>
          author: openHAB Bot <info@openhab.org>
          branch: automated/fetch_openhab-alexa_docs
          delete-branch: true
          title: "[Automated] Fetch openhab-alexa docs"
          labels: automated_pr
```

**Comment by Confectrician** [comment]
I was afraid that we may need to have several versions of a workflow again.

One of my (personal) goals while introducing the actions was to simplify the process and get rid of that many exceptions and custom solutions we currently have with the maven based solution.

See also #1714 for reference.
openhab-addons uses a different folder naming for the bundle docs, than most of the other repositories.

It is always a mess when something doesn't work and you have to think in the special workflow handling for that specific external source.

But for alexa, google, asisstant, mycroft and myopenhab we should be able to get a similar docs structure in place to get some reusable code. So we can simplify it a bit at least.

**Comment by jsetton** [comment]
I just (force) pushed changes that should cover most of the current documentation integrations while providing flexible options. Let me know what you think.

The only outstanding issue is for any integrations that share its base folder with other documentation without using a prefix (e.g. `openhab-android`). The remove existing doc resources step would delete more file than needed. While as you know I am not a fan of using prefixes, does this make sense to move these integrations into separate folders?

**Comment by Confectrician** [comment]
> While as you know I am not a fan of using prefixes, does this make sense to move these integrations into separate folders?

Not really at the current state.
We probalby would have to adapt all scripts for all workflows, where such a file gets processed.
This means scripts in several branches and in the website repository too in several places.

Maybe at a later point, when i am able to start one of my long term goals. => simplyfiy the complete website workflow.
But that's still far away for me.

I will have a look at the lates changes later.

File: .github/workflows/fetch_external_docs_reusable.yml [hunk 1]
```diff
@@ -51,51 +63,54 @@ jobs:
...
+          ref: ${{ inputs.base_target_branch }}
 
       - name: Checkout external repository
         uses: actions/checkout@v2.4.0
         with:
-          repository: ${{inputs.base_source_organization}}/${{inputs.base_source_repository}}
-          path: './external_contents/${{inputs.base_source_repository}}'
+          repository: ${{ inputs.base_source_organization }}/${{ inputs.base_source_repository }}
+          path: ./external_repositories/${{ inputs.base_source_repository }}
+          ref: ${{ inputs.base_source_ref }}
```
[the remark this answers is not in the document]
Comment (jsetton) [reply .github/workflows/fetch_external_docs_reusable.yml:73]: Per the [documentation](https://docs.github.com/en/enterprise-cloud@latest/actions/learn-github-actions/workflow-syntax-for-github-actions#onworkflow_callinputs), if a `default` parameter is not set, the default value of the input is `false` for a boolean, `0` for a number, and `""` for a string.

Also if `ref` is an empty string, the checkout action reverts to using the default branch of the given repository. It's actually better than hardcoding the default value to `main` as it was previously done.

Here is a [success test workflow](https://github.com/jsetton/openhab-docs/actions/runs/1628343624) for these changes that ran on my repository.

**Comment by jsetton** [comment]
> Not really at the current state.
We probalby would have to adapt all scripts for all workflows, where such a file gets processed.
This means scripts in several branches and in the website repository too in several places.

So I just (force) pushed a change that now gives the ability to control if the existing doc should be removed based on a new self-contained flag (defaults to `true`) or the doc file name prefix being defined.

I think this should cover all the different documentation integrations. The only thing I am wondering is if we should have `doc_base_dir` default to empty string opposed to currently `/docs`. I have been going back and worth on this one.

**Comment by Confectrician** [comment]
>The only thing I am wondering is if we should have doc_base_dir default to empty string opposed to currently /docs. I have been going back and forth on this one.

I would leave it with `/docs` for now.
That is the folder we should prefer anyway.

**Review by Confectrician — APPROVED** [review]
Thanks for the work and the huge efforts, to get this as standartized as possible @jsetton 🙂

**Comment by jsetton** [comment]
Once this PR is merged. The `fetch_docs_openhabian` workflow will need to be updated unless you want me to include it in this one.

Basically, the `image_base_name` parameter should be removed since `doc_base_name` now covers all resources.

**Comment by Confectrician** [comment]
Would be nice if you would add it directly.
For me there is no need to do this in an extra pr and we have all related changes to get everything working in one place then.

**Comment by jsetton** [comment]
> Would be nice if you would add it directly.

Done and renamed PR accordingly.

**Comment by jsetton** [comment]
Now that I think about it. I might as well add the other workflows.

**Comment by jsetton** [comment]
And done. Let me know if there are other workflows that should be added.

**Comment by Confectrician** [comment]
Those were the ones i had in mind too for now. Thanks again for the effort. :)

**Review by Confectrician — APPROVED** [review]

**Pull request merged**